clientd/src/lib.rs

```diff
@@ -114,15 +114,144 @@ pub async fn run_dispatcher(
 
 async fn handle_rpc_request(
     request: RpcRequest,
-    fedimint_client: &Client<UserClientConfig>,
+    fedimint_client: Option<&FedimintClient>,
 ) -> Result<HandlerResult, CallError> {
     match request {
         RpcRequest::HealthCheck() => Ok(HandlerResult::HealthCheck()),
-        RpcRequest::Info() => rpc::info(fedimint_client).await,
+        RpcRequest::Info() => {
+            rpc::info(
+                fedimint_client
+                    .expect("fedimint client was passed to the dispatcher")
+                    .as_ref(),
+            )
+            .await
+        }
     }
 }
 
 fn load_from_file<T: DeserializeOwned>(path: &Path) -> T {
     let file = std::fs::File::open(path).expect("Can't read cfg file.");
     serde_json::from_reader(file).expect("Could not parse cfg file.")
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::time::Duration;
+
+    #[test]
+    fn event_subscriber_initialization_works() {
+        // this tests if there is "something" for that event NOT if that something works
+        // for subscribing to events (this will be another test)
+        let all_events = EventKey::iter(); // expect strum works correctly
+        let subscribers = EventSubscribers::default();
+        for event in all_events {
+            assert!(subscribers.as_ref().get(&event).is_some())
+        }
+    }
+
+    #[tokio::test]
+    async fn event_subscribers_can_be_used_to_send_and_recv() {
+        // might be redundant because of the `dispatcher_handles_event` test but its just a small test anyway
+        let subscribers = EventSubscribers::default();
+
+        let mut rx = subscribers
+            .as_ref()
+            .get(&EventKey::Mock)
+            .unwrap()
+            .subscribe();
+        let tx2 = subscribers.as_ref().get(&EventKey::Mock).unwrap().clone();
+
+        // the spawn might be unnecessary but it shows in what context this can be used in
+        tokio::spawn(async move {
+            for _ in 0..5 {
+                tx2.send(Event {
+                    event_key: EventKey::Mock,
+                    data: "test".to_string(),
+                })
+                .unwrap();
+            }
+        })
+        .await
+        .unwrap();
+
+        // FIXME: refactor this to be more clever
+        let mut results = vec![];
+        for _ in 0..5 {
+            if let Ok(msg) = rx.recv().await {
+                results.push(msg);
+            }
+        }
+
+        assert_eq!(
+            results.into_iter().map(|e| e.data).collect::<Vec<String>>(),
+            vec!["test".to_string(); 5]
+        )
+    }
+
+    #[tokio::test]
+    async fn dispatcher_handles_rpc() {
+        let subscribers = EventSubscribers::default();
+        let (dispatcher_tx, dispachter_rx) = mpsc::channel(8);
+        let dispatcher_tx = Arc::new(dispatcher_tx);
+
+        let dispatcher =
+            run_dispatcher(Arc::clone(&dispatcher_tx), dispachter_rx, subscribers, None);
```

Comment:
I see, that's why the client is optional. Two ideas to improve:

* Create a `Box<dyn IClient>` newtype (probably a bit too advanced)
* Just create a client from fake config?

Reply:
mhm true..didnt think about creating a fake config..also `Box<dyn IClient>` I don't really understand what you mean

Reply:
Yep, even the current version works. If you want to learn about `Box<dyn …>` trait objects: https://doc.rust-lang.org/rust-by-example/trait/dyn.html